synapse/lib/module.py

```diff
@@ -0,0 +1,97 @@
+import synapse.eventbus as s_eventbus
+import synapse.telepath as s_telepath
+
+import synapse.lib.config as s_config
+
+class CoreModule(s_eventbus.EventBus,s_config.Configable):
+    '''
+    The CoreModule base class from which cortex modules must extend.
+
+    This module interface implements helper APIs to facilitate cortex
+    extensions.  It is highly encouraged
+
+    To load a module within a cortex, add it to the list of modules in
+    the cortex config ( mostly likely within your dmon config ) as shown:
+
+    # example cortex config
+    {
+        "modules":[
+            ["foopkg.barmod.modctor", {
+                "foo:opt":10,
+                "bar:opt":"http://www.vertex.link"
+            }]
+        ]
+    }
+
+    Modules may extend the cortex in various ways such as:
+
+        * Implement and enforce data model additions
+        * Enrich properties during node creation / modification
+        * Add "by" handlers and side-pocket indexes to extend queries
+        * Add custom storm/swarm operators to the query language
+        * etc etc etc...
+    '''
+
+    def __init__(self, core, conf):
+        s_eventbus.EventBus.__init__(self)
+        s_config.Configable.__init__(self)
+
+        s_telepath.reqNotProxy(core)
+
+        self.core = core
+
+        self.setConfOpts(conf)
```

Comment:
I'm not currently seeing a place for a module implementer to plumb in opts via addConfDef/addConfDefs.  If you moved setConfOpts() after initCoreModule() then a implementor can set use that to addConfDef values.  Here is a patch which allows that.

```diff --git a/synapse/lib/module.py b/synapse/lib/module.py
index 12613e7..20359b1 100644
--- a/synapse/lib/module.py
+++ b/synapse/lib/module.py
@@ -40,10 +40,10 @@ class CoreModule(s_eventbus.EventBus,s_config.Configable):
 
         self.core = core
 
-        self.setConfOpts(conf)
-
         self.initCoreModule()
 
+        self.setConfOpts(conf)
+
     def initCoreModule(self):
         '''
         Module implementers may over-ride this method to initialize the
diff --git a/synapse/tests/test_cortex.py b/synapse/tests/test_cortex.py
index b5f3470..8e22a8a 100644
--- a/synapse/tests/test_cortex.py
+++ b/synapse/tests/test_cortex.py
@@ -35,6 +35,7 @@ class CoreTestModule(s_module.CoreModule):
             props['inet:ipv4:asn'] = 10
 
         self.onFormNode('inet:ipv4',formipv4)
+        self.addConfDef('foobar', **{'type': 'str', 'defval':'woah', 'doc': 'A option'})
 
 class CortexTest(SynTest):
 
@@ -1512,3 +1513,20 @@ class CortexTest(SynTest):
             self.eq( node[1].get('inet:ipv4:asn'), 10 )
 
         self.true( modu.isfini )
+
+    def test_cortex_module_opts(self):
+        with s_cortex.openurl('ram:///') as core:
+
+            node = core.formTufoByProp('inet:ipv4','1.2.3.4')
+            self.eq( node[1].get('inet:ipv4:asn'), -1 )
+
+            mods = ( ('synapse.tests.test_cortex.CoreTestModule', {'foobar': 'duck_quack'}), )
+            core.setConfOpt('modules', mods)
+
+            # directly access the module so we can confirm it gets fini()
+            modu = core.coremods.get('synapse.tests.test_cortex.CoreTestModule')
+
+            node = core.formTufoByProp('inet:ipv4','1.2.3.5')
+            self.eq( node[1].get('inet:ipv4:asn'), 10 )
+
+        self.true( modu.isfini )
```